bittensor/_neuron/text/core_validator/__init__.py

```diff
@@ -513,39 +569,60 @@ def neuron_stats_update(self, neuron_stats: Dict[int, Dict[str, Any]]):
 
         return responsive_uids, list(neuron_stats.keys())  # responsive_uids, queried_uids
 
-    def calculate_weights(self):
+    def calculate_weights(self, responsive_uids: Set, queried_uids: Set):
         r""" Calculates neuron set-weights from weight_key mapped values. Defines weight_key as the neuron stats key
         used to obtain the mapped stat value (typically a Shapley value) that the final set-weights are calculated from.
         """
 
         weight_key = self.weight_key + '!'  # use zeroing key to penalize non-responsive neurons
-        n_topk_peer_weights = self.subtensor.min_allowed_weights
+
+        # === Randomize UIDs in preferred order (responsive -> queried -> rest) ===
+        min_allowed_weights = self.subtensor.min_allowed_weights
         max_allowed_ratio = self.subtensor.max_allowed_min_max_ratio
 
-        # === Calculate neuron weights ===
-        neuron_weights = torch.zeros_like(self.metagraph.S)  # allow unevaluated UIDs to be selected to meet min topk
+        non_responsive_uids = queried_uids - responsive_uids
+        non_queried_uids = set(range(self.metagraph.n)) - queried_uids
+
+        # random.sample(population, k, *, counts=None): Return a k length list of unique elements chosen from
+        # the population sequence or set. Used for random sampling without replacement (so no uid duplicates expected).
+        preferred_uids = (random.sample(list(responsive_uids), len(responsive_uids)) +
+                          random.sample(list(non_responsive_uids), len(non_responsive_uids)) +
+                          random.sample(list(non_queried_uids), len(non_queried_uids)))  # preferred UID random order
+
+        preferred_uids = torch.LongTensor(preferred_uids)
+
+        # === Populate neuron weights ===
+        neuron_weights = torch.zeros_like(self.metagraph.S)  # allow unevaluated UIDs for min_allowed_weights
 
         for uid in self.neuron_stats:
             if weight_key in self.neuron_stats[uid]:
                 neuron_weights[uid] = torch.tensor([self.neuron_stats[uid][weight_key]])
 
-        # Find the n_topk_peer_weights peers to set weights to.
-        topk_weights, topk_uids = bittensor.unbiased_topk(neuron_weights, k=n_topk_peer_weights)
-        topk_weights = bittensor.utils.weight_utils.normalize_max_multiple(x=topk_weights,
-                                                                           multiple=max_allowed_ratio)
-        return topk_uids, topk_weights
+        # === Filter to non-zero weights ===
+        neuron_weights = neuron_weights[preferred_uids]  # rearrange neuron_weights to match preferred_uids order
+        preferred_uids = preferred_uids[neuron_weights > 0]  # filter to non-zero weights
+        neuron_weights = neuron_weights[neuron_weights > 0]  # filter to non-zero weights
+
+        # === Slice min_allowed_weights UIDs ===
+        sample_uids = preferred_uids[:min_allowed_weights]  # slice to min_allowed_weights
```

Comment:
actually i have always been confused, cause `min_allowed_weight` is just the minimum network requirement, 
so shouldnt we set it to `max(min_allowed_weight, len(responsive_uids))` ?

Reply:
Previously and currently validators always fixed the number of weights set to min_allowed_weight, likely because the demands on subtensor are more predictable then and because validators operate more similarly the consensus is improved.